concourse/client/util.py

```diff
@@ -122,3 +156,163 @@ def jobs_not_triggered(
+        next_retry=next_retry.isoformat(),
+        comment=(
+            "pinned by technical user 'concourse', please do not unpin before "
+            f"{next_retry.isoformat()}"
+        ),
+    )
+    concourse_api.pin_and_comment_resource_version(
+        pipeline_name=resource.pipeline_name(),
+        resource_name=resource.name,
+        resource_version_id=resource_version.id(),
+        comment=json.dumps(dataclasses.asdict(pin_comment)),
+    )
+
+
+@contextmanager
+def ensure_resource_unpinned(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    try:
+        _ensure_resource_unpinned(
+            resource=resource,
+            resource_version=resource_version,
+            concourse_api=concourse_api,
+        )
+        yield
+    finally:
+        concourse_api.unpin_resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+        info(f'successfully unpinned {resource.name=}')
+
+
+def _ensure_resource_unpinned(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    previous_retry_time = None
+    maximal_unpin_checktime = datetime.now() + timedelta(minutes=5)
+
+    while maximal_unpin_checktime - datetime.now() >= 0:
+        cc_resource = concourse_api.resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+        if not cc_resource.is_pinned():
+            return
+
+        try:
+            pin_comment = from_dict(
+                data_class=PinComment,
+                data=json.loads(cc_resource.pin_comment()),
+            )
+            if pin_comment.version == resource_version.version():
+                raise PinningUnnecessaryError("Resource is already pinned for the same version")
+
+            next_retry_time = dateutil.parser.isoparse(pin_comment.next_retry)
+            if previous_retry_time and next_retry_time == previous_retry_time:
+                warning(
+                    f'found resource which was not unpinned '
+                    f'{resource.pipeline_name()=} {resource.name=}. Unpinning resource now'
+                )
+                concourse_api.unpin_resource(
+                    pipeline_name=resource.pipeline_name(),
+                    resource_name=resource.name,
+                )
+                return
+        except json.JSONDecodeError:
+            info(
+                f'Resource comment for {resource.name=} of {resource.pipeline_name()=} '
+                'could not be parsed - most likely pinned by human user'
+            )
+        except DaciteError:
+            info(
+                f'Resource comment for {resource.name=} of {resource.pipeline_name()=} '
+                'could not be instantiated'
+            )
+
+        previous_retry_time = next_retry_time
+        time.sleep((next_retry_time - datetime.now()).seconds + 5)
```

Comment:
add comment as to why we add a margin

Reply:
also, there should be a maximum sleep-time

Reply:
sleep time will be maxed by parameter `wait_until_next_retry_seconds` in function `pin_resource_and_trigger_build`

Reply:
yeah, but you do not check for it here. if the timestamp would e.g. be a year in the future, this code would wait for a year

Reply:
also, just checking for whether or not the timestamp changed is not even required (and adds an unnecesary code path). admittedly, this could be used as an _optimisation_. However, let us please not add optimisations right in the first shot. please let us keep this PR focussed, so we (and in particular me) have a chance of reviewing / understanding it properly. the diff is already too big for my taste, and adding optional optimisations is not helping. add optimisations later (after getting the first version merged), if required.

how about simply going for:

```
def acquire_resource_pin(....):
  while "timeout not exceeded and retries left":
    if "resource is unpinned":
      "pin_it_and_return"
    "read_resource_pinning_comment"
    if "resource_pinning_comment is not "from us" (e.g. from a human)":
      "write warning to log + return"
    else: # "resource_pinning_comment _is_ "from us""
       "check announced release timestamp" (note: this time could / should be published whilst locking)
       if "announced release timestamp is in the past":
         log a warning + unpin + pin + return
       else: # "announced release timestamp is in the future":
        wait for max(max_wait_time, announced_release_timestamp)
        # max_wait_time is merely a "safety guard" to prevent erroneous timestamps from causing DOS
```